pkg/util/shared_action_queue.go

```diff
@@ -0,0 +1,93 @@
+package util
+
+import (
+	"sync"
+)
+
+type ActionContext struct {
+	wg         sync.WaitGroup
+	resultChan chan error
+	isWaiting  bool
+}
+
+type Action = func(*ActionContext)
+type ActionGroup = []Action
+
+type SharedActionQueue struct {
+	sharedActionQueue chan func()
+}
+
+// This is a utility designed to run 'Action's to do some work async but limit
+// by a pre-defined number. This allows multiple components to share the same
+// pool of goroutines or other similar resources (like data streams) easily.
+func NewSharedActionQueue(maxConcurrentDownloads uint32) *SharedActionQueue {
```

Comment:
One important aspect of having queueing for this is that it respects `context.Context`. This queue doesn't seem to do that. This means that workers may get stuck on downloading files for certain build actions, even if the client that requested the build action to be run disappears.

Reply:
Done. It now splits the context and up to the consumer to use `ActionCache.Ctx()`.